### PR TITLE
get monster iid (event actions)

### PR DESCRIPTION
**Pull request by JaskRendix**
follows #2179 

PR updates various event actions and replaces **find_monster_by_id** with **get_monster_by_iid**, the reason is similar, `player.find_monster_by_id ` is a sort of bottleneck, the case of add_tech is the most important, I was obliged to add a last parameter called slug just to make possible to add techs to NPCs monster, now it's better (look at the map events); less player dependent.

Adapted some action to **set_monster_** style, sometimes by introducing logger.error and updating messages

### CODE REVIEW

File: tuxemon/event/actions/add_tech.py [hunk 1]
```diff
@@ -21,39 +24,34 @@ class AddTechAction(EventAction):
...
-            raise ValueError(
-                f"No monster found with instance_id {instance_id}",
-            )
+        if self.variable not in player.game_variables:
+            logger.error(f"Game variable {self.variable} not found")
+            return
+
+        monster_id = uuid.UUID(player.game_variables[self.variable])
+        monster = get_monster_by_iid(self.session, monster_id)
+        assert monster
```
Comment (xirsoi) [inline tuxemon/event/actions/add_tech.py]: Please replace this with a check that logs an error, like you did with the game_variables check on line 47

File: tuxemon/event/actions/change_taste.py [hunk 2]
```diff
@@ -23,51 +24,44 @@ class ChangeTasteAction(EventAction):
...
-            monster = trainer.find_monster_in_storage(monster_id)
-
-        if monster is None:
-            logger.error(
-                f"Could not find monster with instance id {monster_id}"
-            )
-            return
+        monster_id = uuid.UUID(player.game_variables[self.variable])
+        monster = get_monster_by_iid(self.session, monster_id)
+        assert monster
```
Comment (xirsoi) [inline tuxemon/event/actions/change_taste.py]: also here

File: tuxemon/event/actions/remove_monster.py [hunk 3]
```diff
@@ -22,28 +25,28 @@ class RemoveMonsterAction(EventAction):
...
+        if self.variable not in player.game_variables:
+            logger.error(f"Game variable {self.variable} not found")
+            return
 
-        monster = trainer.find_monster_by_id(instance_id)
-        if monster is not None:
-            player.remove_monster(monster)
+        monster_id = uuid.UUID(player.game_variables[self.variable])
+        monster = get_monster_by_iid(self.session, monster_id)
+        assert monster and monster.owner
```
Comment (xirsoi) [inline tuxemon/event/actions/remove_monster.py]: and here

File: tuxemon/event/actions/set_monster_flair.py [hunk 4]
```diff
@@ -43,10 +44,8 @@ def start(self) -> None:
             logger.error(f"Game variable {self.variable} not found")
             return
         monster_id = uuid.UUID(player.game_variables[self.variable])
-        monster = player.find_monster_by_id(monster_id)
-        if monster is None:
-            logger.error("Monster not found in party")
-            return
+        monster = get_monster_by_iid(self.session, monster_id)
+        assert monster
```
Comment (xirsoi) [inline tuxemon/event/actions/set_monster_flair.py]: and here

File: tuxemon/event/actions/set_monster_level.py [hunk 5]
```diff
@@ -47,10 +48,8 @@ def start(self) -> None:
                 logger.error(f"Game variable {self.variable} not found")
                 return
             monster_id = uuid.UUID(player.game_variables[self.variable])
-            monster = player.find_monster_by_id(monster_id)
-            if monster is None:
-                logger.error("Monster not found in party")
-                return
+            monster = get_monster_by_iid(self.session, monster_id)
+            assert monster
```
Comment (xirsoi) [inline tuxemon/event/actions/set_monster_level.py]: and here

File: tuxemon/event/actions/set_monster_plague.py [hunk 6]
```diff
@@ -54,8 +55,6 @@ def start(self) -> None:
                 logger.error(f"Game variable {self.variable} not found")
                 return
             monster_id = uuid.UUID(player.game_variables[self.variable])
-            monster = player.find_monster_by_id(monster_id)
-            if monster is None:
-                logger.error("Monster not found in party")
-                return
+            monster = get_monster_by_iid(self.session, monster_id)
+            assert monster
```
Comment (xirsoi) [inline tuxemon/event/actions/set_monster_plague.py]: and another one

File: tuxemon/event/actions/set_monster_status.py [hunk 7]
```diff
@@ -64,8 +65,6 @@ def start(self) -> None:
                 logger.error(f"Game variable {self.variable} not found")
                 return
             monster_id = uuid.UUID(player.game_variables[self.variable])
-            monster = player.find_monster_by_id(monster_id)
-            if monster is None:
-                logger.error("Monster not found in party")
-                return
+            monster = get_monster_by_iid(self.session, monster_id)
+            assert monster
```
Comment (xirsoi) [inline tuxemon/event/actions/set_monster_status.py]: you know the drill :P

File: tuxemon/event/actions/store_monster.py [hunk 8]
```diff
@@ -22,39 +26,37 @@ class StoreMonsterAction(EventAction):
...
-            raise ValueError(
-                f"No monster found with instance_id {instance_id}",
-            )
+        if self.variable not in player.game_variables:
+            logger.error(f"Game variable {self.variable} not found")
+            return
+
+        monster_id = uuid.UUID(player.game_variables[self.variable])
+        monster = get_monster_by_iid(self.session, monster_id)
+        assert monster and monster.owner
```
Comment (xirsoi) [inline tuxemon/event/actions/store_monster.py]: last one maybe?

File: tuxemon/event/actions/trading.py [hunk 9]
```diff
@@ -21,20 +25,27 @@ class TradingAction(EventAction):
...
     name = "trading"
-    removed: str
+    variable: str
     added: str
 
-    def switch_monster(self, removed: Monster) -> None:
-        slot = self.player.monsters.index(removed)
+    def create_monster(self, removed: Monster) -> None:
+        # retrieves character from monster
+        assert removed.owner
```
Comment (xirsoi) [inline tuxemon/event/actions/trading.py]: similar thing here

File: tuxemon/event/actions/trading.py [hunk 10]
```diff
@@ -43,14 +54,49 @@ def switch_monster(self, removed: Monster) -> None:
...
-        self.player.remove_monster(removed)
-        self.player.add_monster(added, slot)
-        self.player.tuxepedia[self.added] = SeenStatus.caught
+        character.remove_monster(removed)
+        character.add_monster(added, slot)
+        character.tuxepedia[added.slug] = SeenStatus.caught
+
+    def switch_monster(self, removed: Monster, added: Monster) -> None:
+        # retrieves both character from monsters
+        assert removed.owner and added.owner
```
Comment (xirsoi) [inline tuxemon/event/actions/trading.py]: no assert!

File: tuxemon/event/actions/trading.py [hunk 11]
```diff
@@ -43,14 +54,49 @@ def switch_monster(self, removed: Monster) -> None:
...
     def start(self) -> None:
-        self.player = self.session.player
-        iid = uuid.UUID(self.player.game_variables[self.removed])
-        removed = self.player.find_monster_by_id(iid)
-        if removed:
-            self.switch_monster(removed)
+        player = self.session.player
+        _monster_id = uuid.UUID(player.game_variables[self.variable])
+        monster_id = get_monster_by_iid(self.session, _monster_id)
+        assert monster_id
```
Comment (xirsoi) [inline tuxemon/event/actions/trading.py]: and another

**Review by xirsoi — CHANGES_REQUESTED** [review]
please address the comments

**Review by xirsoi — DISMISSED** [review]

**Review by bitcraft — DISMISSED** [review]

**Comment by bitcraft** [comment]
conflicts

**Comment by JaskRendix** [comment]
@bitcraft fixed

**Review by bitcraft — APPROVED** [review]

**Pull request merged**